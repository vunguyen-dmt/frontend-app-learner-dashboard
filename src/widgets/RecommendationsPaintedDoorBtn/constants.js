export const COLLAPSED_NAVBAR = 'collapsedNavbar';
export const EXPANDED_NAVBAR = 'expendedNavbar';
export const RECOMMENDATIONS_PANEL = 'recommendationsPanel';
