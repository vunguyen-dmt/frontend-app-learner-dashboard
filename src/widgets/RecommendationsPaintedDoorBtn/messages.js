import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  recommendationsFeatureText: {
    id: 'RecommendationsPanel.recommendationsFeatureText',
    defaultMessage: 'Personalized recommendations feature is not yet available. We are working hard on bringing it to your learner home in the near future.',
    description: 'recommendations feature text',
  },
  recommendationsAlertText: {
    id: 'RecommendationsPanel.recommendationsAlertText',
    defaultMessage: 'Would you like to be alerted when it becomes available?',
    description: 'recommendations alerted text',
  },
  recommendationsModalHeading: {
    id: 'RecommendationsPanel.recommendationsModalHeading',
    defaultMessage: 'Thank you for your interest!',
    description: 'Heading of modal',
  },
  modalSkipButton: {
    id: 'RecommendationsPanel.modalSkipButton',
    defaultMessage: 'Skip for now',
    description: 'button for Skip for now',
  },
  modalCountMeButton: {
    id: 'RecommendationsPanel.modalCountMeButton',
    defaultMessage: 'Count me in!',
    description: 'button for Count me in!',
  },
  recommendedForYou: {
    id: 'learnerVariantDashboard.recommendedForYou',
    defaultMessage: 'Recommended For You',
    description: 'Header link for recommended page.',
  },
  seeAllRecommendationsButton: {
    id: 'RecommendationsPanel.seeAllRecommendationsButton',
    defaultMessage: 'See All Recommendations',
    description: 'Button to see all recommendations',
  },
});

export default messages;
