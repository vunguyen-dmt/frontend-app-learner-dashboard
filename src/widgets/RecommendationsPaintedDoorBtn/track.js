import { StrictDict } from 'utils';
import { createEventTracker } from 'data/services/segment/utils';

export const eventNames = StrictDict({
  variationGroup: 'edx.bi.user.recommendation_home.variation.group',
  recommendationHomeBtnClicked: 'edx.bi.user.recommendation_home.btn.clicked',
  recommendationHomeModalInterestBtnClicked: 'edx.bi.user.recommendation_home.modal.interest_btn.clicked',
  recommendationHomeModalSkipBtnClicked: 'edx.bi.user.recommendation_home.modal.skip_btn.clicked',
});

export const trackPaintedDoorVariationGroup = (variation) => {
  createEventTracker(eventNames.variationGroup, {
    variation,
    page: 'dashboard',
  })();
};

export const trackPaintedDoorRecommendationHomeBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeBtnClicked, {
    variation,
    page: 'dashboard',
  })();
};

export const trackPaintedDoorRecommendationHomeInterestBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeModalInterestBtnClicked, {
    variation,
  })();
};

export const trackPaintedDoorRecommendationHomeSkipBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeModalSkipBtnClicked, {
    variation,
  })();
};
