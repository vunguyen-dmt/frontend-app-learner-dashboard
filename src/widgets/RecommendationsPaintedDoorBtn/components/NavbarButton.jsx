import React from 'react';
import { Button } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import messages from '../messages';
import { COLLAPSED_NAVBAR, EXPANDED_NAVBAR } from '../constants';

export const NavbarButton = ({ placement, handleClick }) => {
  const { formatMessage } = useIntl();

  return (
    <Button
      as="a"
      className={classNames({
        'p-4': placement === EXPANDED_NAVBAR,
      })}
      variant="inverse-primary"
      onClick={handleClick}
    >
      {formatMessage(messages.recommendedForYou)}
    </Button>
  );
};

NavbarButton.propTypes = {
  placement: PropTypes.oneOf([COLLAPSED_NAVBAR, EXPANDED_NAVBAR]).isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default NavbarButton;
