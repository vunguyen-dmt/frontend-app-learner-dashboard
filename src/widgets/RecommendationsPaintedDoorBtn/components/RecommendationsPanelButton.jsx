import React from 'react';
import { Button } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import PropTypes from 'prop-types';
import messages from '../messages';

export const RecommendationsPanelButton = ({ handleClick }) => {
  const { formatMessage } = useIntl();

  return (
    <Button
      as="a"
      variant="brand"
      onClick={handleClick}
    >
      {formatMessage(messages.seeAllRecommendationsButton)}
    </Button>
  );
};

RecommendationsPanelButton.propTypes = {
  handleClick: PropTypes.func.isRequired,
};

export default RecommendationsPanelButton;
