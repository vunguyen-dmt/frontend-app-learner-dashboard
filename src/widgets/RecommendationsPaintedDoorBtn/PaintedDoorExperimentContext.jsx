import React from 'react';
import PropTypes from 'prop-types';
import { StrictDict } from 'utils';
import * as module from './PaintedDoorExperimentContext';
import {
  useEmailConfirmationData,
  useHasAvailableDashboards,
  useRequestIsPending,
} from '../../data/redux/hooks';
import { RequestKeys } from '../../data/constants/requests';
import { trackPaintedDoorVariationGroup } from './track';

export const state = StrictDict({
  enterpriseUser: (val) => React.useState(val), // eslint-disable-line
  experimentData: (val) => React.useState(val), // eslint-disable-line
});

const PAINTED_DOOR_RECOMMENDATIONS_EXP_ID = 25116810832;
const PAINTED_DOOR_RECOMMENDATIONS_PAGE = 'url_targeting_for_van1604_recommendations_painted_door_exp';
const PAINTED_DOOR_RECS_EXP_NAVBAR_BTN_VARIATION = 'btn_navbar';
const PAINTED_DOOR_RECS_EXP_WIDGET_BTN_VARIATION = 'btn_widget';
const PAINTED_DOOR_RECS_EXP_CONTROL_VARIATION = 'control';

export function getPaintedDoorRecommendationsExperimentVariation() {
  try {
    if (window.optimizely && window.optimizely.get('data').experiments[PAINTED_DOOR_RECOMMENDATIONS_EXP_ID]) {
      const selectedVariant = window.optimizely.get('state').getVariationMap()[PAINTED_DOOR_RECOMMENDATIONS_EXP_ID];
      return selectedVariant?.name;
    }
  } catch (e) { /* empty */ }
  return '';
}

export function activatePaintedDoorRecommendationsExperiment() {
  window.optimizely = window.optimizely || [];
  window.optimizely.push({
    type: 'page',
    pageName: PAINTED_DOOR_RECOMMENDATIONS_PAGE,
  });
}

export const useIsEnterpriseUser = () => {
  const [enterpriseUser, setEnterpriseUser] = module.state.enterpriseUser({
    isEnterpriseUser: false,
    isLoading: true,
  });

  const initIsPending = useRequestIsPending(RequestKeys.initialize);
  const hasAvailableDashboards = useHasAvailableDashboards();
  const confirmationData = useEmailConfirmationData();

  React.useEffect(() => {
    if (!initIsPending && Object.keys(confirmationData).length && hasAvailableDashboards) {
      setEnterpriseUser(prev => ({
        ...prev,
        isEnterpriseUser: true,
        isLoading: false,
      }));
    } else if (!initIsPending && Object.keys(confirmationData).length && !hasAvailableDashboards) {
      setEnterpriseUser(prev => ({
        ...prev,
        isEnterpriseUser: false,
        isLoading: false,
      }));
    }
  }, [initIsPending]); // eslint-disable-line react-hooks/exhaustive-deps

  return enterpriseUser;
};

export const PaintedDoorExperimentContext = React.createContext({
  experimentVariation: null,
  isPaintedDoorNavbarBtnVariation: null,
  isPaintedDoorWidgetBtnVariation: null,
  isPaintedDoorControlVariation: null,
  experimentLoading: null,
});

export const PaintedDoorExperimentProvider = ({ children }) => {
  const [experimentData, setExperimentData] = module.state.experimentData({
    experimentVariation: '',
    isPaintedDoorNavbarBtnVariation: false,
    isPaintedDoorWidgetBtnVariation: false,
    isPaintedDoorControlVariation: false,
    experimentLoading: true,
  });
  const enterpriseUser = useIsEnterpriseUser();

  const contextValue = React.useMemo(
    () => ({
      ...experimentData,
    }),
    [experimentData],
  );

  React.useEffect(() => {
    let timer = null;
    if (!enterpriseUser.isLoading && !enterpriseUser.isEnterpriseUser) {
      activatePaintedDoorRecommendationsExperiment();
      timer = setTimeout(() => {
        const variation = getPaintedDoorRecommendationsExperimentVariation();
        setExperimentData(prev => ({
          ...prev,
          experimentVariation: variation,
          isPaintedDoorNavbarBtnVariation: variation === PAINTED_DOOR_RECS_EXP_NAVBAR_BTN_VARIATION,
          isPaintedDoorWidgetBtnVariation: variation === PAINTED_DOOR_RECS_EXP_WIDGET_BTN_VARIATION,
          isPaintedDoorControlVariation: variation === PAINTED_DOOR_RECS_EXP_CONTROL_VARIATION,
          experimentLoading: false,
        }));
        trackPaintedDoorVariationGroup(variation);
      }, 500);
    }
    return () => clearTimeout(timer);
  }, [enterpriseUser]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <PaintedDoorExperimentContext.Provider value={contextValue}>
      {children}
    </PaintedDoorExperimentContext.Provider>
  );
};

export const usePaintedDoorExperimentContext = () => React.useContext(PaintedDoorExperimentContext);

PaintedDoorExperimentProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default PaintedDoorExperimentProvider;
