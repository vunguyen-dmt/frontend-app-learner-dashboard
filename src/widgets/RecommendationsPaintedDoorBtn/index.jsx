import React from 'react';
import PropTypes from 'prop-types';
import NavbarButton from './components/NavbarButton';
import PaintedDoorModal from './components/PaintedDoorModal';
import { usePaintedDoorModal } from './components/PaintedDoorModal/hooks';
import { COLLAPSED_NAVBAR, EXPANDED_NAVBAR, RECOMMENDATIONS_PANEL } from './constants';
import RecommendationsPanelButton from './components/RecommendationsPanelButton';
import { trackPaintedDoorRecommendationHomeBtnClicked } from './track';

export const RecommendationsPaintedDoorBtn = ({ placement, experimentVariation }) => {
  const { isModalOpen, toggleModal } = usePaintedDoorModal();

  const handleClick = () => {
    toggleModal();
    trackPaintedDoorRecommendationHomeBtnClicked(experimentVariation);
  };

  const getPaintedDoorButton = () => {
    if ([COLLAPSED_NAVBAR, EXPANDED_NAVBAR].includes(placement)) {
      return (
        <NavbarButton handleClick={handleClick} placement={placement} />
      );
    } if (placement === RECOMMENDATIONS_PANEL) {
      return (
        <RecommendationsPanelButton handleClick={handleClick} />
      );
    }
    return null;
  };

  return (
    <>
      {getPaintedDoorButton()}
      <PaintedDoorModal isOpen={isModalOpen} onClose={toggleModal} variation={experimentVariation} />
    </>
  );
};

RecommendationsPaintedDoorBtn.propTypes = {
  placement: PropTypes.oneOf([COLLAPSED_NAVBAR, EXPANDED_NAVBAR, RECOMMENDATIONS_PANEL]).isRequired,
  experimentVariation: PropTypes.string.isRequired,
};

export default RecommendationsPaintedDoorBtn;
