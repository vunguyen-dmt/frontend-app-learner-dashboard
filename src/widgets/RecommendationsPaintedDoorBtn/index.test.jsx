import React from 'react';
import { mount, shallow } from 'enzyme';
import { Button, ModalDialog } from '@edx/paragon';
import RecommendationsPaintedDoorBtn from './index';
import { EXPANDED_NAVBAR, RECOMMENDATIONS_PANEL } from './constants';
import NavbarButton from './components/NavbarButton';
import RecommendationsPanelButton from './components/RecommendationsPanelButton';
import { trackPaintedDoorRecommendationHomeBtnClicked } from './track';

jest.mock('./components/PaintedDoorModal/hooks', () => ({
  usePaintedDoorModal: jest.fn(() => ({
    isModalOpen: false,
    toggleModal: jest.fn(),
  })),
}));

jest.mock('./track', () => ({
  trackPaintedDoorRecommendationHomeBtnClicked: jest.fn(),
}));

describe('RecommendationsPaintedDoorBtn', () => {
  let props = {
    placement: RECOMMENDATIONS_PANEL,
    experimentVariation: '',
  };

  it('matches snapshot', () => {
    expect(shallow(<RecommendationsPaintedDoorBtn {...props} />)).toMatchSnapshot();
  });

  it('renders painted door modal', () => {
    const wrapper = shallow(<RecommendationsPaintedDoorBtn {...props} />);

    expect(wrapper.find(ModalDialog)).toBeTruthy();
  });

  it('renders painted door navbar button', () => {
    props = {
      ...props,
      placement: EXPANDED_NAVBAR,
    };
    const wrapper = shallow(<RecommendationsPaintedDoorBtn {...props} />);

    expect(wrapper.find(NavbarButton).exists()).toBe(true);
    expect(wrapper.find(RecommendationsPanelButton).exists()).toBe(false);
  });

  it('renders painted door recommendations panel button', () => {
    props = {
      ...props,
      placement: RECOMMENDATIONS_PANEL,
    };
    const wrapper = shallow(<RecommendationsPaintedDoorBtn {...props} />);

    expect(wrapper.find(NavbarButton).exists()).toBe(false);
    expect(wrapper.find(RecommendationsPanelButton).exists()).toBe(true);
  });

  it('test no button (null) rendered for invalid placement', () => {
    props = {
      ...props,
      placement: '',
    };
    const wrapper = shallow(<RecommendationsPaintedDoorBtn {...props} />);

    expect(wrapper.find(NavbarButton).exists()).toBe(false);
    expect(wrapper.find(RecommendationsPanelButton).exists()).toBe(false);
  });

  it('test track event is fired on navbar button click', () => {
    props = {
      ...props,
      placement: EXPANDED_NAVBAR,
    };
    const wrapper = mount(<RecommendationsPaintedDoorBtn {...props} />);
    const navbarButton = wrapper.find(NavbarButton);

    navbarButton.find(Button).simulate('click');

    expect(trackPaintedDoorRecommendationHomeBtnClicked).toHaveBeenCalled();
  });

  it('test track event is fired on recommendations panel button click', () => {
    props = {
      ...props,
      placement: RECOMMENDATIONS_PANEL,
    };
    const wrapper = mount(<RecommendationsPaintedDoorBtn {...props} />);
    const navbarButton = wrapper.find(RecommendationsPanelButton);

    navbarButton.find(Button).simulate('click');

    expect(trackPaintedDoorRecommendationHomeBtnClicked).toHaveBeenCalled();
  });
});
