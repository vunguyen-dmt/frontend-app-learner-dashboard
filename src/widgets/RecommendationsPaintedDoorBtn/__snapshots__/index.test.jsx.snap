// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationsPaintedDoorBtn matches snapshot 1`] = `
<Fragment>
  <RecommendationsPanelButton
    handleClick={[Function]}
  />
  <ModalView
    isOpen={false}
    onClose={[MockFunction]}
    variation=""
  />
</Fragment>
`;
